### PR TITLE
[MRG] Fix for #814

### PR DESCRIPTION
**Pull request by jayantj**
The current code attempts to decode all characters for vocab words individually as ascii.
Instead, the code is changed to do the following - characters are read as raw bytes first, then decoded as utf-8 once the entire word has been read.

Also includes a test fasttext model with non-ascii vocab words, and checks vector lookup for the same.

Haven't been able to test the newly released FastText vectors for wiki in various languages since the models are large-ish (~10 GB). 

Related issue reported [here](https://github.com/RaRe-Technologies/gensim/issues/814#issuecomment-283381412)

### CODE REVIEW

File: gensim/test/test_data/non_ascii_fasttext.vec [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+byl -0.89449 1.4131 
+tu -1.1255 1.5505 
+tedy -1.1693 1.6446 
+teto -1.2134 1.546 
+bylo -0.86091 1.3805 
+kde -1.3468 1.7507 
+ke -1.0699 1.6688 
+pravé -0.9391 1.5172 
+nad -1.3404 1.7661 
+nejsou -0.85023 1.5033
```
Comment (anmol01gulati) [inline gensim/test/test_data/non_ascii_fasttext.vec]: Shouldn't this file be in a binary format in the package and later converted to text? This also works, but just feel that would be better.

**Comment by anmol01gulati** [comment]
@jayantj This looks good, but should we wait for to test this on the new pre-trained models first and then ask Lev to merge? Though I feel, it shouldn't create any difference, as you have already added a test for non-ascii words.

**Comment by tmylk** [comment]
Agree with @anmol01gulati. Does it work on https://github.com/facebookresearch/fastText/blob/master/pretrained-vectors.md, @jayantj  ?

**Comment by piskvorky** [comment]
Does fastText always output utf8? Is the encoding really fixed?

**Comment by piskvorky** [comment]
@jayantj @tmylk  the fastText file header is incorrect (states I wrote the code in 2013). Please fix.

**Comment by piskvorky** [comment]
@tmylk use better issue/PR titles, so search engines can index them and users find them.

**Comment by piskvorky** [comment]
@jayantj status on the questions above?

**Comment by jayantj** [comment]
The fastText readme says that the training file should contain utf-8 encoded text. That was the rationale behind choosing `utf8` as the encoding while deserializing the model.

Looking at the fastText code and experimenting with different encodings though, fastText doesn't really decode the bytes in the training file and simply treats the input file as a stream of bytes.

The only place it does make use of the values inside the bytes is while tokenizing the text, where it checks each byte against whitespace characters (`' ', '\n', '\r', '\t', '\v', '\f'`)

So, encodings which have the same values for these characters as ascii would produce sane output. (`utf16` and `utf32` don't, lots of others do). Original encoding information isn't stored anywhere though.

I can think of two ways to handle this - 

1. Assume words in input file are utf8 encoded (since FastText readme mentions it, and it might be enforced in the future), and raise a more friendly error when the input isn't valid utf8
2. Provide an parameter `encoding` allowing the user to specify the input file encoding (defaulting to `utf8`, of course)

**Comment by jayantj** [comment]
Re: file header, I'll create a single PR with the updated header and the fix for the above issue, whichever one we go with.

**Comment by tmylk** [comment]
Option 2 sounds as the most flexible

**Pull request merged**